Improve suggestions for `needless_range_loop`

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->
Hey there. First of all: Great work to all the devs. Clippy is a real handy tool when developing in Rust.

Now about my proposal: the `needless_range_loop` gives pretty good suggestions even for non default ranges. But if a range loop like the following is linted, one could suggest to use a mutable iterator using `iter_mut()`:
``` Rust
for i in 1..arr.len()-1 {
    arr[i] = <some_value>;
}
```
Clippy would suggest something like `for <item> in arr.iter().take(arr.len()-1).skip(1) {` but in this case `iter_mut()` must be used to assign new values to `<item>`. I don't know if there is an easy way to detect cases like this one but I think it would be a great improvement.

for the first case, clippy now suggest something that doesn't compile:
```rust
fn main() {
    let mut arr = vec![1, 2, 3];
    for i in 1..arr.len()-1 {
        arr[i] = 6;
    }
}
```

```
warning: the loop variable `i` is only used to index `arr`.
 --> src/main.rs:4:14
  |
4 |     for i in 1..arr.len()-1 {
  |              ^^^^^^^^^^^^^^
  |
  = note: #[warn(needless_range_loop)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.205/index.html#needless_range_loop
help: consider using an iterator
  |
4 |     for <item> in arr.iter_mut().take(arr.len()-1).skip(1) {
  |
```

applying the suggestion:
```rust
fn main() {
    let mut arr = vec![1, 2, 3];
    for item in arr.iter_mut().take(arr.len()-1).skip(1) {
        *item = 6
    }
}
```

```
error[E0502]: cannot borrow `arr` as immutable because it is also borrowed as mutable
 --> src/main.rs:4:37
  |
4 |     for item in arr.iter_mut().take(arr.len()-1).skip(1) {
  |                 ---                 ^^^                - mutable borrow ends here
  |                 |                   |
  |                 |                   immutable borrow occurs here
  |                 mutable borrow occurs here

error: aborting due to previous error
```

--- a/clippy_lints/src/loops.rs
+++ b/clippy_lints/src/loops.rs
@@ -3,7 +3,7 @@ use crate::utils::paths;
 use crate::utils::sugg::Sugg;
 use crate::utils::usage::{is_unused, mutated_variables};
 use crate::utils::{
-    get_enclosing_block, get_parent_expr, get_trait_def_id, has_iter_method, higher, implements_trait,
+    contains_name, get_enclosing_block, get_parent_expr, get_trait_def_id, has_iter_method, higher, implements_trait,
     is_integer_const, is_no_std_crate, is_refutable, is_type_diagnostic_item, last_path_segment, match_trait_method,
     match_type, match_var, multispan_sugg, qpath_res, snippet, snippet_opt, snippet_with_applicability,
     snippet_with_macro_callsite, span_lint, span_lint_and_help, span_lint_and_sugg, span_lint_and_then, sugg,
@@ -1276,6 +1276,8 @@ fn check_for_loop_range<'tcx>(
 
                 let skip = if starts_at_zero {
                     String::new()
+                } else if visitor.indexed_mut.contains(&indexed) && contains_name(indexed, start) {
+                    return;
                 } else {
                     format!(".skip({})", snippet(cx, start.span, ".."))
                 };
@@ -1302,6 +1304,8 @@ fn check_for_loop_range<'tcx>(
 
                     if is_len_call(end, indexed) || is_end_eq_array_len(cx, end, limits, indexed_ty) {
                         String::new()
+                    } else if visitor.indexed_mut.contains(&indexed) && contains_name(indexed, take_expr) {
+                        return;
                     } else {
                         match limits {
                             ast::RangeLimits::Closed => {

--- a/tests/ui/needless_range_loop2.rs
+++ b/tests/ui/needless_range_loop2.rs
@@ -82,6 +82,20 @@ fn main() {
     for i in 1..3 {
         println!("{}", arr[i]);
     }
+
+    // Fix #5945
+    let mut vec = vec![1, 2, 3, 4];
+    for i in 0..vec.len() - 1 {
+        vec[i] += 1;
+    }
+    let mut vec = vec![1, 2, 3, 4];
+    for i in vec.len() - 3..vec.len() {
+        vec[i] += 1;
+    }
+    let mut vec = vec![1, 2, 3, 4];
+    for i in vec.len() - 3..vec.len() - 1 {
+        vec[i] += 1;
+    }
 }
 
 mod issue2277 {

